# Baseline Models (Naïve and Seasonal Naïve)

Create two simple forecasts for benzina and gasolio:

- **Naïve model**: next week’s price = this week’s price.
- **Seasonal naïve** model: next week’s price = price from the same week last year (52 weeks ago).

Then compute error metrics (MAE, RMSE, MAPE).
These numbers will be your baseline.

In [1]:
import numpy as np
import pandas as pd
from sklearn.metrics import mean_absolute_error, mean_squared_error

weekly = pd.read_csv('../data/processed/weekly_dataset.csv', parse_dates=['date'])

weekly = weekly.sort_values('date').reset_index(drop=True)
weekly = weekly.set_index('date')

Naïve forecast

In [2]:
for target in ['benzina', 'gasolio']:
    y_true = weekly[target].iloc[1:]                  # real values (from 2nd week onward)
    y_pred = weekly[target].shift(1).iloc[1:]         # previous week’s value

    mae = mean_absolute_error(y_true, y_pred)
    rmse = np.sqrt(mean_squared_error(y_true, y_pred))
    mape = np.mean(np.abs((y_true - y_pred) / y_true)) * 100

    print(f'{target.upper()} — Naïve model:')
    print(f'MAE: {mae:.4f}, RMSE: {rmse:.4f}, MAPE: {mape:.2f}%\n')


BENZINA — Naïve model:
MAE: 0.0111, RMSE: 0.0198, MAPE: 0.72%

GASOLIO — Naïve model:
MAE: 0.0113, RMSE: 0.0214, MAPE: 0.77%



Seasonal naïve (1 year = 52 weeks)

In [3]:
for target in ['benzina', 'gasolio']:
    y_true = weekly[target].iloc[52:]
    y_pred = weekly[target].shift(52).iloc[52:]

    mae = mean_absolute_error(y_true, y_pred)
    rmse = np.sqrt(mean_squared_error(y_true, y_pred))
    mape = np.mean(np.abs((y_true - y_pred) / y_true)) * 100

    print(f'{target.upper()} — Seasonal naïve (52 weeks):')
    print(f'MAE: {mae:.4f}, RMSE: {rmse:.4f}, MAPE: {mape:.2f}%\n')


BENZINA — Seasonal naïve (52 weeks):
MAE: 0.1304, RMSE: 0.1587, MAPE: 8.41%

GASOLIO — Seasonal naïve (52 weeks):
MAE: 0.1541, RMSE: 0.1890, MAPE: 10.74%



The **naïve model** performed surprisingly well, with an average error of about 0.011 €/L for both benzina and gasolio. This means that fuel prices in Italy tend to change very gradually from week to week. The **seasonal naïve model** performed much worse, confirming that yearly seasonality is not a strong predictor for weekly fuel prices.